### DISCUSSION TITLE
Localized defaultValue, label(s)

### DISCUSSION THREAD

**Discussion by chladog** [payloadcms/payload#497] [Feature Requests & Ideas]
Currently there is no option to pass defaultValue of the field based on the locale.
E.g. I have locales: ```en```,  ```cs``` and I need to provide different for each:
Something like:
```JS
            {
              name: "ceremonyLocation",
              type: "textarea",
              label: "Ceremony Location",
              localized: true,
              defaultValue: {
                en: `The wedding ceremony will take place at `,
                cs: `Obřad se bude konat `,
              },
```

if particular locale not present then should use default locale.
the object of locales could be optional and setting simple string should be also allowed (not localized default value)
```JS
defaultValue?: string | {
     [key: locale]: value
}
```

Same applies for label(s) properties that could be i18n this way.

**Comment by jmikrut** [comment]
Hey @chladog — great idea. We can add this to our planned features. 

Question - should we consider making `defaultValue` accept a function so that you could return a value based on other fields, locale, etc?

Would that be more powerful?

**Reply by chladog** [reply]
Yes, the defaultValue factory would be perfect.  
Then we could create dynamic vals, like prefilling user's data in the contact form on server side (currently I do that in FE).

**Reply by chladog** [reply]
Just thinking about this statement ```you could return a value based on other fields``` <- would be possible but only for pre-saved fields values (defaultValue wouldn't regenerate in UI real-time), therefore for defaultValue based on sibling-data the FE implementation is imho preferable. Though the example I wrote is still quite practical ^ ```defaultValue based on data of **different document** related to current user``` is very possible and much leaner to implement server-side than FE (as client-app would have to load both docs first and comes with quite bulky implementation for such trivial thing).

To make this feature complete the same **value factory** would also be great for other fields like ```admin.description``` and additionally maybe introduce```admin.placeholder```, or at least some ```metadata: JSON``` field that can generate and hold case-specific values (e.g. schema for custom component based on structure of other fields/docs).

_I had ```options```  on my mind as well, that could generate options based on sibling fields (or other docs) server-side which would be great for strong-validation. Though I'm not sure how this could be handled in GrahpQL where they are strong-typed enums - maybe in such case they could just be String - but in the other way if we get mentioned "metadata" we can implement own Select component on "type: text" (and kind of Select type with "Other" user-typed value)_

**Reply by DanRibbens** [reply]
`admin.description` already supports functions. You can see an example in the [docs](https://payloadcms.com/docs/fields/overview#description)

`admin.placeholder` could do the same thing, though my gut says there isn't much demand to do so.

The `metadata` idea is interesting, but I think you could achieve the same functionality by adding a sibling field configured for your purpose.

Making `options` dynamic would have more implications beyond GraphQL. We'd need to change validation, database enums and probably more.

I always appreciate your feedback @chladog, keep it coming!

**Reply by chladog** [reply]
I guess the need depends on Payload use-case. In my case the app itself is basically **front-end form app built up from the Fields[]** definitions, I use Payload's admin area for managing users, mailing, user entries, additional hidden fields etc.
Currently I write extended field definitions with "metadata: { ... }" properties that I remove before passing it to Payload config. It works great for static data, but it's unnecessary hard to implement any kind conditioned and dynamic data (like this topic's **i18n defaultValue**, **i18n labels**, placeholders, icon names, custom input data, etc.

As you mention I could add additional _fields_ for holding such metadata and run some logic in hooks, however my forms are huge and it's not comfortable to mix user-entered form fields with system fields, schema gets bigger, also frontend application gets complex as I can't simply loop and render all form fields, I need to filter out system fields, pair system fields with corrent form fields merge the data and implement dynamic logic on client-side - all that with i18n data in all supported languages.

**Reply by DanRibbens** [reply]
Thanks for the extra context.

I had thought you could use `admin.hidden` for system fields and maybe prefix field names with something like `sys` or `_` for filtering later. I didn't realize you'd want to add this for every field, I get why you wouldn't want to now.

If you want to propose details for your metadata feature idea we could see consider adding it or building it as a plugin.

**Reply by DanRibbens** [reply]
I think we're close to having this one complete.

https://github.com/payloadcms/payload/pull/547

**Reply by DanRibbens** [reply]
Released in `v0.16.1`.

I used your example @chladog to write the [docs](https://payloadcms.com/docs/fields/overview#default-values), thanks for that!

**Reply by chladog** [reply]
Hello Dan, this is perfect!
Just a little fix - the example sets     
```JS
admin: {
      description: ({ user, locale }) => (`${translation[locale]} ${user.name}`)
    }
```
instead of ```defaultValue```.


I'm about to test this feature soon. I hope you will implement this for ```label(s)``` as well ( once #515 tackled )
Keep up the great work!

**Comment by bratvanov** [comment]
As in #273, this feature request can be closed, since `defaultValue` as a function has been supported for a while:
https://payloadcms.com/docs/fields/overview#default-values

> Here is an example of a `defaultValue` function:
> ```ts
> import type { Field } from 'payload'
> 
> const translation: {
>   en: 'Written by'
>   es: 'Escrito por'
> }
> 
> export const myField: Field = {
>   name: 'attribution',
>   type: 'text',
>   defaultValue: ({ user, locale, req }) =>
>     `${translation[locale]} ${user.name}`,
> }
> ```
>
> Tip: You can use async `defaultValue` functions to fill fields with data from API requests or Local API using `req.payload`.
